ui: properly segregate/aggregate app sql stats

couchand opened the pull request:
Previously we had been a bit lazy about the apps that statement statistics roll into, but feedback highlighted that the app was an important distinction for users.  This adds support for apps to the statements list and details pages.

The statements list now has a filter above the list to choose an app:
<img width="548" alt="screen shot 2018-06-25 at 1 04 56 pm" src="https://user-images.githubusercontent.com/793969/41864643-aeba45bc-7878-11e8-85ff-60d39d71beb6.png">

The details page now lists all apps a query appears in (when viewing all apps) or just the stats relevant to a single app.
<img width="466" alt="screen shot 2018-06-25 at 1 05 05 pm" src="https://user-images.githubusercontent.com/793969/41864691-d196db7c-7878-11e8-89d5-32717753a935.png">

Fixes: #26990

couchand committed:
ui: filter by app on statements page

This adds a filter to the statements page to view statement statistics
for a single app or across all apps.

Release note (admin ui change): Statement statistics can now be filtered
by app on the Statements page.

couchand committed:
ui: properly aggregate statement stats

Release note: None

couchand committed:
ui: handle app separation on details page

Release note: None

couchand committed:
ui: helper methods to flatten and combine statement stats

Release note: None

couchand committed:
ui: really aggregate statistics correctly

Release note: None

couchand committed:
ui: really handle aggregation on details page

Release note: None

couchand committed:
ui: only show (unset) app if any statement stats have no app

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/26949)
<!-- Reviewable:end -->

couchand commented:
The code and git history are admittedly a bit messy, I want to keep this moving along so I've opened the PR in the current form, happy to take any and all feedback @vilterp.

@knz can you take a look at this and see if it's getting us closer to what you'd expect here?  Thanks!

vilterp commented:
Saw a runtime error while running TPCC:

http://localhost:8080/#/statement/(unset)/SELECT%20s_quantity%2C%20s_ytd%2C%20s_order_cnt%2C%20s_remote_cnt%2C%20s_data%2C%20s_dist_10%20FROM%20stock%20WHERE%20(s_i_id%2C%20s_w_id)%20IN%20((_%2C%20_)%2C%20(_%2C%20_)%2C%20(_%2C%20_)%2C%20(_%2C%20_)%2C%20(_%2C%20_)%2C%20(_%2C%20_)%2C%20(_%2C%20_)%2C%20(_%2C%20_)%2C%20(_%2C%20_))%20ORDER%20BY%20s_i_id

![✂-2](https://files.reviewableusercontent.io/images/reviewable/7341/3tszaIqFOqwU4AdnMSDIDvG1.png)

![✂-1](https://files.reviewableusercontent.io/images/reviewable/7341/Nrv49juAaif56DmkGI-5NCke.png)

---

Review status: :shipit: complete! 0 of 0 LGTMs obtained

---



*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/26949#-:-LFsv0q-4HuWMKMPgN-d:bfjekxl)*
<!-- Sent from Reviewable.io -->

knz commented:
> can you take a look at this and see if it's getting us closer to what you'd expect here?

I looked at the screenshot and I confirm it is getting us closer to what I'd expect (and what customers would expect).

I will test more once the PR lands.

---

Reviewed 1 of 1 files at r1.
Review status: :shipit: complete! 0 of 0 LGTMs obtained

---



*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/26949#-:-LFwA1E22XWD66BK5hIU:b-7tgkeu)*
<!-- Sent from Reviewable.io -->

couchand committed:
ui: fix npe on statement stats aggregation

Previously we were looking for a statement in the app literally
named '(unset)', but that's of course not what's meant by the
filter.

Release note: None

couchand committed:
ui: move route for statement within an app

This move makes the URL segment parsing consistent: if the first
segment is /statement, the next segment is an app name.  On the
other hand, if the first segment is /statements, the next segment
is a query fingerprint.

Release note: None

vilterp commented:
Review status: :shipit: complete! 0 of 0 LGTMs obtained

---

*[pkg/ui/src/util/appStats.ts, line 35 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/26949#-LFx1xBd4ugEF2MKwae9:-LFx1xBd4ugEF2MKwaeA:b-24gmvp) ([raw file](https://github.com/cockroachdb/cockroach/blob/0c82703bb5175758bcfa28bce150df7988ddd528/pkg/ui/src/util/appStats.ts#L35)):*
> ```XML
> }
> 
> export function addStatementStats(a: StatementStatistics, b: StatementStatistics) {
> ```

maybe add a comment that this is duplicated from `app_stats.go`

---

*[pkg/ui/src/views/statements/statementsPage.tsx, line 214 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/26949#-LFx2CKl97nZLFnbYGGd:-LFx2CKl97nZLFnbYGGe:bvwx10w) ([raw file](https://github.com/cockroachdb/cockroach/blob/0c82703bb5175758bcfa28bce150df7988ddd528/pkg/ui/src/views/statements/statementsPage.tsx#L214)):*
> ```XML
> }
> 
> const selectStatements = createSelector(
> ```

There's some nontrivial logic in these selectors; would be nice to have a couple of tests. E.g. send in a couple of queries with the same fingerprint but different `failed` or `distsql`; see that they get coalesced correctly.

If not tests demonstrating what this code is doing, please at least add some basic comments above each selector, e.g. "`selectApps` iterates over all fingerprints, returning the set of apps they use" and "`selectStatements` coalesces the returned StatementStats based on fingerprint, not `(fingerprint, distsql, failed)`, as they are originally grouped.". Otherwise, it's pretty hard to understand what's going on here.

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/26949)*
<!-- Sent from Reviewable.io -->

couchand committed:
ui: clarify mean/dev bar chart on statement details page

By shrinking the wings drawn for the standard deviation, the mean is
highlighted, making reading the chart much easier.

Release note (admin ui change): Improve the readability of the mean and
std. dev. bar chart on the statement details page.

couchand committed:
ui: refactor css on statement details page

Release note: None

couchand committed:
ui: add bar for overall latency on statment details

Release note: None

couchand committed:
ui: add std dev wings to latency chart on statement list

Release note (admin ui change): Add a visualization of the standard
deviation of the latency of statements on the statements list.

couchand committed:
ui: clarify the copy on the statement stats list

Release note: None

couchand committed:
ui: put extended statement stats help text in tooltip

Release note: None

couchand committed:
ui: tooltipify titles on statement details

Release note: None

couchand committed:
ui: tooltipify title text on statements list

Release note: None

couchand committed:
ui: cleanup titles on bar charts

Release note: None

couchand committed:
ui: add std dev wings to rows chart on list page

Release note: None

couchand committed:
ui: add comment on ported app stats code

Release note: None

couchand committed:
ui: rename totalStatements to totalFingerprints

To help keep things clear.  Also add some comments.

Release note: None

couchand committed:
ui: extract protobuf/long init for reuse elsewhere

Release note: None

couchand committed:
ui: add tests for statements page selectors

Release note: None

couchand committed:
ui: add tests for statement details selector

Release note: None

couchand committed:
ui: correct display of NaN std dev

Previously, if a statement had only been executed once, it would get a NaN
when calculating standard deviation due to a divide-by-zero.  This fixes it
for charting purposes by setting the std dev to zero in this case.

Fixes: #26990
Release note (admin ui change): Fixes a bug where standard deviations of
statement statistics could cause a charting error.

couchand commented:
Review status: :shipit: complete! 0 of 0 LGTMs obtained

---

*[pkg/ui/src/util/appStats.ts, line 35 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/26949#-LFx1xBd4ugEF2MKwae9:-LG19eMQ6YlWhkyZm0lZ:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/0c82703bb5175758bcfa28bce150df7988ddd528/pkg/ui/src/util/appStats.ts#L35)):*
<details><summary><i>Previously, vilterp (Pete Vilter) wrote…</i></summary><blockquote>

maybe add a comment that this is duplicated from `app_stats.go`
</blockquote></details>

Done.

---

*[pkg/ui/src/views/statements/statementsPage.tsx, line 214 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/26949#-LFx2CKl97nZLFnbYGGd:-LG19gc-7xylJDfbmxkW:b-hem8hc) ([raw file](https://github.com/cockroachdb/cockroach/blob/0c82703bb5175758bcfa28bce150df7988ddd528/pkg/ui/src/views/statements/statementsPage.tsx#L214)):*
<details><summary><i>Previously, vilterp (Pete Vilter) wrote…</i></summary><blockquote>

There's some nontrivial logic in these selectors; would be nice to have a couple of tests. E.g. send in a couple of queries with the same fingerprint but different `failed` or `distsql`; see that they get coalesced correctly.

If not tests demonstrating what this code is doing, please at least add some basic comments above each selector, e.g. "`selectApps` iterates over all fingerprints, returning the set of apps they use" and "`selectStatements` coalesces the returned StatementStats based on fingerprint, not `(fingerprint, distsql, failed)`, as they are originally grouped.". Otherwise, it's pretty hard to understand what's going on here.
</blockquote></details>

Good point.  Added short comments and what seem to be reasonably-complete tests.

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/26949)*
<!-- Sent from Reviewable.io -->

couchand committed:
ui: use tooltip component for full statement on list page

Release note: None

vilterp reviewed (approved):
Nice tests & tooltips!

Quick Q: is there supposed to be a little red fleck on the right of the count bar on the table page? 
![image](https://user-images.githubusercontent.com/7341/42052601-3384b168-7add-11e8-9158-4bcdd4b60bc2.png)

Can't wait to get this on the left nav. I'd say it's about ready, once we have cross-node aggregation!

vilterp commented:
Suggestion, not blocking this PR: the tooltip for getting the full statement is good, but I still wonder if it would be better to just `overflow: ellipse` that cell. I'd like to be able to e.g. open up the statements page for TPCC and cmd+f to see if there are any joins in it. (Still might miss them if ellipsified, but seems better than the summaries).

vilterp commented:
I'm skeptical that we'll be able to find some kind of structured summary that's better than the fingerprint itself, ellipsified so that all the rows are the same height. But open to ideas / brainstorming.

vilterp commented:
Noice! 
![image](https://user-images.githubusercontent.com/7341/42062542-d5ab99b6-7afb-11e8-913d-b98a9e54bd26.png)
